Deferrable operators guide (#304)
astronomer/airflow-guides: guides/deferrable-operators.md

date: 2021-12-25T00:00:00.000Z
slug: "deferrable-operators"
heroImagePath: null
tags: ["Operators", "Concurrency", "Resources", "Sensors", "Workers"]
---

## Overview

Prior to Airflow 2.2, all task execution occurred within your worker resources. For tasks whose work was occurring outside of Airflow (e.g. a Spark Job), your tasks would sit idle waiting for a success or failure signal. These idle tasks would occupy worker slots for their entire duration, potentially queuing other tasks and delaying their start times.

With the release of Airflow 2.2, Airflow has introduced a new way to run tasks in your environment: Deferrable operators. These operators leverage Python's [asyncio](https://docs.python.org/3/library/asyncio.html) library to efficiently run tasks waiting for an external resource to finish. This frees up your workers, allowing you to utilize those resources more effectively. In this guide, we'll walk through the concepts of deferrable operators, as well as the new components introduced to Airflow related to this feature.

## Deferrable Operator Concepts

There are some terms and concepts that are important to understand when discussing deferrable operators:

> - **asyncio:** This is the Python [library](https://docs.python.org/3/library/asyncio.html) that is used as a foundation for multiple asynchronous frameworks. This library is core to Deferrable Operators functionality, and is used when writing triggers.

Reviewer: ```suggestion
- **[asyncio](https://docs.python.org/3/library/asyncio.html):** This Python library is used as a foundation for multiple asynchronous frameworks. This library is core to deferrable operator's functionality, and is used when writing triggers.
```

> - **asyncio:** This is the Python [library](https://docs.python.org/3/library/asyncio.html) that is used as a foundation for multiple asynchronous frameworks. This library is core to Deferrable Operators functionality, and is used when writing triggers.

Reviewer: ```suggestion
- **[asyncio](https://docs.python.org/3/library/asyncio.html):** This Python library is used as a foundation for multiple asynchronous frameworks. This library is core to deferrable operator's functionality, and is used when writing triggers.
```

> - **asyncio:** This is the Python [library](https://docs.python.org/3/library/asyncio.html) that is used as a foundation for multiple asynchronous frameworks. This library is core to Deferrable Operators functionality, and is used when writing triggers.

Reviewer: Resolving and editing in a new suggestion.

- **Triggers:** These are small, asynchronous pieces of Python code. Due to their asynchronous nature, they coexist efficiently in a single process known as the triggerer.

> - **Triggerer:** This is a new airflow service (like a scheduler or a worker) that runs an [asyncio event loop](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio-event-loop) in your Airflow environment. Running a triggerer component is essential for using deferrable operators. Depending on the available resources and the workload of your triggers, you can run hundreds to thousands of triggers in a single triggerer process.

Reviewer: ```suggestion
- **Triggerer:** This is a new airflow service (like a scheduler or a worker) that runs an [asyncio event loop](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio-event-loop) in your Airflow environment. Running a triggerer is essential for using deferrable operators. Depending on the available resources and the workload of your triggers, you can run hundreds to thousands of triggers in a single triggerer process.
```

> - **Deferred:** This is a new Airflow task state (medium purple color) introduced to indicate that a task has paused its execution, released the worker slot, and has submitted a trigger to be picked up by the triggerer process to continue the task’s execution.

Reviewer: I agree, I removed the `to continue the task's execution` piece, as this is all discussed throughout the doc.

> With traditional operators, a task might poll an external system while waiting for a process to complete. Even though the task might not be doing significant work, it would still occupy a worker slot. This can prevent other tasks from running. Visually, that looks something like the diagram below:

Reviewer: ```suggestion
With traditional operators, a task might submit a job to an external system (e.g. a Spark Cluster), and then poll the status of that job until it completes. Even though the task might not be doing significant work, it would still occupy a worker slot during the polling process. As worker slots become occupied, tasks may be queued resulting in delayed start times. Visually, this is represented in the diagram below:
```

![Classic Worker](https://assets2.astronomer.io/main/guides/deferrable-operators/classic_worker_process.png)

> With deferrable operators, worker slots can be released while a task is waiting for an external system. The task is then deferred (suspended), and polling is offloaded to the triggerer, which can run many asynchronous polling tasks concurrently in a single process. When the terminal status for the job is received, the task resumes, taking a worker slot while it finishes. Visually, that looks like this updated diagram:

Reviewer: ```suggestion
With traditional operators, a task might submit a job to an external system (e.g. a Spark Cluster), and then poll the status of that job until it completes. Even though the task might not be doing significant work, it would still occupy a worker slot during the polling process. As worker slots become occupied, tasks may be queued resulting in delayed start times. Visually, this is represented in the diagram below:
```

> With deferrable operators, worker slots can be released while a task is waiting for an external system. The task is then deferred (suspended), and polling is offloaded to the triggerer, which can run many asynchronous polling tasks concurrently in a single process. When the terminal status for the job is received, the task resumes, taking a worker slot while it finishes. Visually, that looks like this updated diagram:

Reviewer: ```suggestion
With deferrable operators, worker slots can be released while polling for job status. When the task is deferred (suspended), the polling process is offloaded to the triggerer, freeing up the worker slot. The triggerer can run many asynchronous polling tasks concurrently in a single process, preventing this work from occupying your worker resources. When the terminal status for the job is received, the task resumes, taking a worker slot while it finishes. Visually, this is represented in the diagram below:
```

> With deferrable operators, worker slots can be released while a task is waiting for an external system. The task is then deferred (suspended), and polling is offloaded to the triggerer, which can run many asynchronous polling tasks concurrently in a single process. When the terminal status for the job is received, the task resumes, taking a worker slot while it finishes. Conceptually, that looks like this updated diagram:

Reviewer 2: ```suggestion
With deferrable operators, worker slots can be released while a task is waiting for an external system. The task is then deferred (suspended), and polling is offloaded to the triggerer, which can run many asynchronous polling tasks concurrently in a single process. When the terminal status for the job is received, the task resumes, taking a worker slot while it finishes. Visually, that looks like this updated diagram:
```

![Deferrable Worker](https://assets2.astronomer.io/main/guides/deferrable-operators/deferrable_operator_process.png)

> ## When and Why to Use Deferrable Operators

Reviewer 2: ```suggestion
## When and Why To Use Deferrable Operators
```

> In general, deferrable operators should be used whenever you have long-running tasks that might occupy a worker slot for longer than is necessary and you want to reduce Airflow’s resource utilization. For example, using deferrable operators in sensor tasks might reduce your operational costs. In particular, if you are currently working with smart sensors to poll an external job or system, deferrable operators will be a preferable, more stable solution and better supported by Airflow in the long term.

Reviewer: ```suggestion
In general, deferrable operators should be used whenever you have tasks that occupy a worker slot while polling for a condition in an external system. For example, using deferrable operators for sensor tasks (e.g. poking for a file on an SFTP server) may result in efficiency gains and reduced operational costs. In particular, if you are currently working with [smart sensors](https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html#smart-sensors), you should consider using deferrable operators for these tasks. They will be a preferable and more flexible solution, better supported by Airflow in the long term.
```

> In general, deferrable operators should be used whenever you have long-running tasks that might occupy a worker slot for longer than is necessary and you want to reduce Airflow’s resource utilization. For example, if you have many sensor tasks using deferrable operators may reduce costs. In particular, if you are currently working with Smart Sensors to poll an external job or system, deferrable operators will be a preferable, more stable solution and better supported by Airflow in the long term.

Reviewer 2: ```suggestion
In general, deferrable operators should be used whenever you have long-running tasks that might occupy a worker slot for longer than is necessary and you want to reduce Airflow’s resource utilization. For example, using deferrable operators in sensor tasks might reduce your operational costs. In particular, if you are currently working with smart sensors to poll an external job or system, deferrable operators will be a preferable, more stable solution and better supported by Airflow in the long term.
```

> Currently, the following operators are available in Airflow as deferrable:

Reviewer 2: ```suggestion
Currently, the following deferrable operators are available in Airflow:
```

- [TimeSensorAsync](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/sensors/time_sensor/index.html?highlight=timesensor#module-contents)

> - [DateTimeSensorAsync](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/sensors/date_time/index.html#airflow.sensors.date_time.DateTimeSensorAsync)

Reviewer 2: Maybe worth a cheeky note explaining that "async" here means the same thing as deferrable, and that these words used interchangeably in Airflow? I can imagine folks getting confused with that (as I once was).

Author: Good call, I added a note in the concepts section above

> However, this list will grow quickly as the Airflow community makes more investments into these operators. In the meantime, you can also create your own (more on this in the last section). Additionally, Astronomer maintains some deferrable operators [available only on Astronomer Runtime](https://docs.astronomer.io/cloud/deferrable-operators/#astronomers-deferrable-operators).

Reviewer 2: ```suggestion
However, this list will grow quickly as the Airflow community makes more investments into these operators. In the meantime, you can also create your own (more on this in the last section of this guide). Additionally, Astronomer maintains some deferrable operators [available only on Astronomer Runtime](https://docs.astronomer.io/cloud/deferrable-operators/#astronomers-deferrable-operators).
```

There are numerous benefits to using deferrable operators. Some of the most notable are:

> - **Reduced Resource Consumption:** Depending on the workload, one triggerer can run hundreds of deferred tasks concurrently. This can lead to a reduction in the number of workers needed to run tasks during periods of high concurrency. With less workers needed, you are able to scale down the underlying infrastructure of your Airflow environment.

Reviewer 2: ```suggestion
- **Reduced resource consumption:** Depending on the workload, one triggerer can run hundreds of deferred tasks concurrently. This can lead to a reduction in the number of workers needed to run tasks during periods of high concurrency. With less workers needed, you are able to scale down the underlying infrastructure of your Airflow environment.
```

> - **Resiliency Against Restarts:** Triggers are stateless by design. This means your deferred tasks will not be set to a failure state if a triggerer needs to be restarted due to a deployment or infrastructure issue. They will resume once a triggerer is up and running in your environment.

Reviewer 2: ```suggestion
- **Resiliency against restarts:** Triggers are stateless by design. This means your deferred tasks will not be set to a failure state if a triggerer needs to be restarted due to a deployment or infrastructure issue. Once a triggerer is back up and running in your environment, your deferred tasks will resume.
```

> - **Paves the Way to Event-Based DAGs:** The presence of `asyncio` in core Airflow is seen as the basis for event-triggered DAGs.

Reviewer 2: ```suggestion
- **Paves the way to event-based DAGs:** The presence of `asyncio` in core Airflow is a valuable foundation for event-triggered DAGs.
```

Reviewer: Updated `valuable` to `potential`.

## Example Workflow Using Deferrable Operators

> Let's say we have a DAG that is scheduled to run a sensor every minute, and where each task can take up to 20 minutes. Using the default settings with 1 worker, we can see that after 20 minutes we have 16 tasks running, each holding a worker slot:

Reviewer 2: ```suggestion
Let's say we have a DAG that is scheduled to run a sensor every minute, where each task can take up to 20 minutes. Using the default settings with 1 worker, we can see that after 20 minutes we have 16 tasks running, each holding a worker slot:
```

![Classic Tree View](https://assets2.astronomer.io/main/guides/deferrable-operators/classic_tree_view.png)

> As worker slots are held during task execution time, we would need at least 20 worker slots available for this DAG to ensure future runs are not delayed. In order to increase concurrency, we would need to add additional resources to our Airflow infrastructure (e.g. another worker pod). 

Reviewer 2: ```suggestion
Because worker slots are held during task execution time, we would need at least 20 worker slots available for this DAG to ensure that future runs are not delayed. To increase concurrency, we would need to add additional resources to our Airflow infrastructure (e.g. another worker pod). 
```

> With Airflow 2.2, we can leverage a deferrable operator to offload this idle work from the Workers. This will release the worker slot from sensing for job status, so it can complete other work across your Airflow environment. With our updated DAG below, we see that all 20 tasks have entered a state of deferred, indicating that the sensing jobs will now be run by triggers in the triggerer process. This process is designed to handle these types of jobs at scale, with the potential of 100's of sensing processes running concurrently on a single triggerer.

Reviewer: ```suggestion
If we leverage a deferrable operator for this sensor, we are able to achieve full concurrency, while allowing our worker to complete additional work across your Airflow environment. With our updated DAG below, we see that all 20 tasks have entered a state of deferred, indicating that these sensing jobs will now be run concurrently by triggers in the triggerer process.
```

> With Airflow 2.2, we can leverage a deferrable operator to offload this idle work from the Workers. This will release the worker slot from sensing for job status, so it can complete other work across your Airflow environment. With our updated DAG below, we see that all 20 tasks have entered a state of deferred, indicating that the sensing jobs will now be run by triggers in the triggerer process. This process is designed to handle these types of jobs at scale, with the potential of 100's of sensing processes running concurrently on a single triggerer.

Reviewer 2: ```suggestion
With Airflow 2.2, we can use a deferrable operator to offload this idle work from the workers. This releases the worker slot from sensing for job status so that it can complete other work across your Airflow environment. With our updated DAG below, we see that all 20 tasks have entered a state of deferred, indicating that the sensing jobs are now completed by triggers in the triggerer process. This process is designed to handle these types of jobs at scale, with the potential for 100's of sensing processes running concurrently on a single triggerer.
```

Reviewer: Refactoring this paragraph as it is essentially a repeat of the paragraphs above.

   start_date=datetime(2021, 12, 22, 20, 0),
   end_date=datetime(2021, 12, 22, 20, 19),
   schedule_interval="* * * * *",
   catchup=True,
   max_active_runs=32,
   max_active_tasks=32
) as dag:
 
   async_sensor = DateTimeSensorAsync(
       task_id="async_task",
       target_time="""{{ macros.datetime.utcnow() + macros.timedelta(minutes=20) }}""",
       pool="async",
   )

from datetime import datetime
from airflow import DAG
from airflow.sensors.date_time import DateTimeSensor
 
with DAG(
   "sync_dag",
   start_date=datetime(2021, 12, 22, 20, 0),
   end_date=datetime(2021, 12, 22, 20, 19),
   schedule_interval="* * * * *",
   catchup=True,
   max_active_runs=32,
   max_active_tasks=32
) as dag:
 
   sync_sensor = DateTimeSensor(
       task_id="sync_task",
       target_time="""{{ macros.datetime.utcnow() + macros.timedelta(minutes=20) }}""",
       pool="sync",
   )
```

## Running Deferrable Tasks in your Airflow Environment

To start a triggerer process, run `airflow triggerer` in your Airflow environment. You should see an output similar to the below image.

![Triggerer Logs](https://assets2.astronomer.io/main/guides/deferrable-operators/triggerer_logs.png)

> Note that if you are using [Astronomer Cloud](https://docs.astronomer.io/cloud/deferrable-operators#prerequisites), the triggerer will be run for you automatically if you are on Astronomer Runtime 4.0+. If you are using Astronomer Enterprise, you can add a triggerer to your Airflow deployment from the deployment settings tab.

Reviewer: Note that if you are using [Astronomer Cloud](https://docs.astronomer.io/cloud/deferrable-operators#prerequisites), the triggerer runs automatically if you are on Astronomer Runtime 4.0+. If you are using Astronomer Enterprise 0.26+, you can add a triggerer to an Airflow 2.2+ deployment in the **Deployment Settings** tab. This [guide](https://docs.astronomer.io/enterprise/configure-deployment#triggerer) details the steps for configuring this feature in the platform.

> Note that if you are using [Astronomer Cloud](https://docs.astronomer.io/cloud/deferrable-operators#prerequisites), the triggerer will be run for you automatically if you are on Astronomer Runtime 4.0+. If you are using Astronomer Enterprise, you can add a triggerer to your Airflow deployment from the deployment settings tab.

Reviewer 2: ```suggestion
Note that if you are using [Astronomer Cloud](https://docs.astronomer.io/cloud/deferrable-operators#prerequisites), the triggerer runs automatically if you are on Astronomer Runtime 4.0+. If you are using Astronomer Enterprise, you can add a triggerer to your Airflow deployment from the **Deployment Settings** tab.
```

Reviewer: @jwitz Should we include our docs for configuring the triggerer on Enterprise?

Note that if you are using [Astronomer Cloud](https://docs.astronomer.io/cloud/deferrable-operators#prerequisites), the triggerer runs automatically if you are on Astronomer Runtime 4.0+. If you are using Astronomer Enterprise, you can add a triggerer to your Airflow deployment from the **Deployment Settings** tab. This [guide](https://docs.astronomer.io/enterprise/configure-deployment#triggerer) details the steps for configuring this feature in the Platform.

Reviewer: Resolving and updating in a new change.

> As tasks are raised into a deferred state, triggers will be registered in the triggerer. You are able to set the number of concurrent triggers that can be run in a single triggerer process with the [`default_capacity`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#triggerer) configuration setting in Airflow. This can also be set via the environment variable `AIRFLOW__TRIGGERER__DEFAULT_CAPACITY`. By default, this is set to 1,000.

Reviewer 2: ```suggestion
As tasks are raised into a deferred state, triggers are registered in the triggerer. You are able to set the number of concurrent triggers that can be run in a single triggerer process with the [`default_capacity`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#triggerer) configuration setting in Airflow. This can also be set via the `AIRFLOW__TRIGGERER__DEFAULT_CAPACITY` environment variable. By default, this variable's value is `1,000`.
```

### High Availability

> Note that triggers are designed to be highly-available. You can implement this by starting multiple triggerer processes. Similar to the [HA scheduler](https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html#running-more-than-one-scheduler) introduced in Airflow 2.0, Airflow will ensure they co-exist with correct locking and HA. You can reference the [Airflow docs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html#high-availability) for further reading on this topic.

Reviewer 2: ```suggestion
Note that triggers are designed to be highly-available. You can implement this by starting multiple triggerer processes. Similar to the [HA scheduler](https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html#running-more-than-one-scheduler) introduced in Airflow 2.0, Airflow ensures that they co-exist with correct locking and HA. You can reference the [Airflow docs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html#high-availability) for further information on this topic.
```

---
title: "Deferrable Operators"
description: "How to implement deferrable operators to save cost and resources with Airflow."
date: 2021-12-25T00:00:00.000Z
slug: "deferrable-operators"
heroImagePath: null
tags: ["Operators", "Concurrency", "Resources", "Sensors", "Workers"]
---

## Overview

Prior to Airflow 2.2, all task execution occurred within your worker resources. For tasks whose work was occurring outside of Airflow (e.g. a Spark Job), your tasks would sit idle waiting for a success or failure signal. These idle tasks would occupy worker slots for their entire duration, potentially queuing other tasks and delaying their start times.

> With the release of Airflow 2.2, Airflow has introduced a new way to run tasks in your environment: Deferrable operators. These operators leverage Python's [asyncio](https://docs.python.org/3/library/asyncio.html) library to efficiently run tasks waiting for an external resource to finish. This frees up your workers, allowing you to utilize those resources more effectively. In this guide, we'll walk through the concepts of deferrable operators, as well as the new components introduced to Airflow related to this feature.

Reviewer: ```suggestion
With the release of Airflow 2.2, Airflow has introduced a new way to run tasks in your environment: deferrable operators. These operators leverage Python's [asyncio](https://docs.python.org/3/library/asyncio.html) library to efficiently run tasks waiting for an external resource to finish. This frees up your workers, allowing you to utilize those resources more effectively. In this guide, we'll walk through the concepts of deferrable operators, as well as the new components introduced to Airflow related to this feature.
```

Reviewer: Just a case change.

## Deferrable Operator Concepts

There are some terms and concepts that are important to understand when discussing deferrable operators:

- **[asyncio](https://docs.python.org/3/library/asyncio.html):** This Python library is used as a foundation for multiple asynchronous frameworks. This library is core to deferrable operator's functionality, and is used when writing triggers.
- **Triggers:** These are small, asynchronous pieces of Python code. Due to their asynchronous nature, they coexist efficiently in a single process known as the triggerer.

> - **Triggerer:** This is a new airflow service (like a scheduler or a worker) that runs an [asyncio event loop](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio-event-loop) in your Airflow environment. Running a triggerer is essential for using deferrable operators. Depending on the available resources and the workload of your triggers, you can run hundreds to thousands of triggers in a single triggerer process.

Reviewer: ```suggestion
- **Triggerer:** This is a new airflow service (like a scheduler or a worker) that runs an [asyncio event loop](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio-event-loop) in your Airflow environment. Running a triggerer is essential for using deferrable operators.
```

Reviewer: Removing this as it is redundant and covered in the `Reduced Resource Consumption` section below.

- **Deferred:** This is a new Airflow task state (medium purple color) introduced to indicate that a task has paused its execution, released the worker slot, and submitted a trigger to be picked up by the triggerer process.

> Note: The terms "deferrable" and "async" or "asynchronous" are often used interchangeably. They mean the same thing in this context.

> With traditional operators, a task might submit a job to an external system (e.g. a Spark Cluster), and then poll the status of that job until it completes. Even though the task might not be doing significant work, it would still occupy a worker slot during the polling process. As worker slots become occupied, tasks may be queued resulting in delayed start times. Visually, this is represented in the diagram below:

Reviewer 2: ```suggestion
With traditional operators, a task might submit a job to an external system (e.g. a Spark Cluster) and then poll the status of that job until it completes. Even though the task might not be doing significant work, it would still occupy a worker slot during the polling process. As worker slots become occupied, tasks may be queued resulting in delayed start times. Visually, this is represented in the diagram below:
```

![Classic Worker](https://assets2.astronomer.io/main/guides/deferrable-operators/classic_worker_process.png)

With deferrable operators, worker slots can be released while polling for job status. When the task is deferred (suspended), the polling process is offloaded as a trigger to the triggerer, freeing up the worker slot. The triggerer has the potential to run many asynchronous polling tasks concurrently, preventing this work from occupying your worker resources. When the terminal status for the job is received, the task resumes, taking a worker slot while it finishes. Visually, this is represented in the diagram below:

![Deferrable Worker](https://assets2.astronomer.io/main/guides/deferrable-operators/deferrable_operator_process.png)

## When and Why To Use Deferrable Operators

> In general, deferrable operators should be used whenever you have tasks that occupy a worker slot while polling for a condition in an external system. For example, using deferrable operators for sensor tasks (e.g. poking for a file on an SFTP server) may result in efficiency gains and reduced operational costs. In particular, if you are currently working with [smart sensors](https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html#smart-sensors), you should consider using deferrable operators for these tasks. They will be a preferable and more flexible solution, better supported by Airflow in the long term.

Reviewer 2: ```suggestion
In general, deferrable operators should be used whenever you have tasks that occupy a worker slot while polling for a condition in an external system. For example, using deferrable operators for sensor tasks (e.g. poking for a file on an SFTP server) can result in efficiency gains and reduced operational costs. In particular, if you are currently working with [smart sensors](https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html#smart-sensors), you should consider using deferrable operators for these tasks. They will be a preferable and more flexible solution, better supported by Airflow in the long term.
```

Currently, the following deferrable operators are available in Airflow:

- [TimeSensorAsync](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/sensors/time_sensor/index.html?highlight=timesensor#module-contents)
- [DateTimeSensorAsync](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/sensors/date_time/index.html#airflow.sensors.date_time.DateTimeSensorAsync)

However, this list will grow quickly as the Airflow community makes more investments into these operators. In the meantime, you can also create your own (more on this in the last section of this guide). Additionally, Astronomer maintains some deferrable operators [available only on Astronomer Runtime](https://docs.astronomer.io/cloud/deferrable-operators/#astronomers-deferrable-operators).

There are numerous benefits to using deferrable operators. Some of the most notable are:

> - **Reduced resource consumption:** Depending on the workload, one triggerer can run hundreds of deferred tasks concurrently. This can lead to a reduction in the number of workers needed to run tasks during periods of high concurrency. With less workers needed, you are able to scale down the underlying infrastructure of your Airflow environment.

Reviewer: ```suggestion
- **Reduced resource consumption:** Depending on the available resources and the workload of your triggers, you can run hundreds to thousands of triggers in a single triggerer process. This can lead to a reduction in the number of workers needed to run tasks during periods of high concurrency. With less workers needed, you are able to scale down the underlying infrastructure of your Airflow environment.
```

Reviewer: Consolidated this information under this section.

 
with DAG(
   "sync_dag",
   start_date=datetime(2021, 12, 22, 20, 0),
   end_date=datetime(2021, 12, 22, 20, 19),
   schedule_interval="* * * * *",
   catchup=True,
   max_active_runs=32,
   max_active_tasks=32
) as dag:
 
   sync_sensor = DateTimeSensor(
       task_id="sync_task",
       target_time="""{{ macros.datetime.utcnow() + macros.timedelta(minutes=20) }}""",
       pool="sync",
   )
```

## Running Deferrable Tasks in your Airflow Environment

To start a triggerer process, run `airflow triggerer` in your Airflow environment. You should see an output similar to the below image.

![Triggerer Logs](https://assets2.astronomer.io/main/guides/deferrable-operators/triggerer_logs.png)

Note that if you are using [Astronomer Cloud](https://docs.astronomer.io/cloud/deferrable-operators#prerequisites), the triggerer runs automatically if you are on Astronomer Runtime 4.0+. If you are using Astronomer Enterprise 0.26+, you can add a triggerer to an Airflow 2.2+ deployment in the **Deployment Settings** tab. This [guide](https://docs.astronomer.io/enterprise/configure-deployment#triggerer) details the steps for configuring this feature in the platform.

> As tasks are raised into a deferred state, triggers are registered in the triggerer. You are able to set the number of concurrent triggers that can be run in a single triggerer process with the [`default_capacity`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#triggerer) configuration setting in Airflow. This can also be set via the `AIRFLOW__TRIGGERER__DEFAULT_CAPACITY` environment variable. By default, this variable's value is `1,000`.

Reviewer 2: ```suggestion
As tasks are raised into a deferred state, triggers are registered in the triggerer. You can set the number of concurrent triggers that can run in a single triggerer process with the [`default_capacity`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#triggerer) configuration setting in Airflow. This can also be set via the `AIRFLOW__TRIGGERER__DEFAULT_CAPACITY` environment variable. By default, this variable's value is `1,000`.
```
